controllers/policies/policyserver_controller.go

```diff
@@ -152,71 +192,43 @@ func (r *PolicyServerReconciler) SetupWithManager(mgr ctrl.Manager) error {
-			// Updates since they will invalidate the newever versions of the
-			// object.
-			policy, ok := object.(*policiesv1alpha2.ClusterAdmissionPolicy)
-			if !ok {
-				r.Log.Error(err, "object is not type of ClusterAdmissionPolicy: %#v", policy)
-				return []ctrl.Request{}
-			}
-
-			if policy.Status.PolicyStatus == "" {
-				policy.Status.PolicyStatus = policiesv1alpha2.ClusterAdmissionPolicyStatusUnscheduled
-				err := r.Reconciler.UpdateAdmissionPolicyStatus(context.Background(), policy)
-				if err != nil {
-					r.Log.Error(err, "cannot update status of policy "+policy.Name)
-					return []ctrl.Request{}
+		Watches(&source.Kind{Type: &policiesv1alpha2.ClusterAdmissionPolicy{}}, handler.Funcs{
+			CreateFunc: func(e event.CreateEvent, queue workqueue.RateLimitingInterface) {
+				queue.Add(watchClusterAdmissionPolicy(r, e.Object))
+			},
+			UpdateFunc: func(e event.UpdateEvent, queue workqueue.RateLimitingInterface) {
+				newPolicy, ok := e.ObjectNew.(*policiesv1alpha2.ClusterAdmissionPolicy)
+				if !ok {
+					r.Log.Error(err, "ObjectNew is not type of ClusterAdmissionPolicy: %#v", newPolicy)
+					queue.Add(ctrl.Request{})
+					return
 				}
-			}
-
-			var policyServers policiesv1alpha2.PolicyServerList
-			err := r.List(context.Background(), &policyServers, client.MatchingFields{constants.PolicyServerIndexName: policy.Spec.PolicyServer})
-			if err != nil {
-				r.Log.Error(err, "cannot list PolicyServers corresponding to policy "+policy.Name)
-				return []ctrl.Request{}
-			}
-			if len(policyServers.Items) == 0 {
-				return r.reconcileOrphanPolicies(policy)
-			}
-
-			if policy.DeletionTimestamp != nil {
-				// policy scheduled for deletion hence read-only, can't change
-				// status, reconcile loop removes Finalizer:
-				return []ctrl.Request{
-					{
-						NamespacedName: client.ObjectKey{
-							Name: policy.Spec.PolicyServer,
-						},
-					},
+				queue.Add(watchClusterAdmissionPolicy(r, newPolicy))
+			},
+			DeleteFunc: func(e event.DeleteEvent, queue workqueue.RateLimitingInterface) {
+				policy, ok := e.Object.(*policiesv1alpha2.ClusterAdmissionPolicy)
+				if !ok {
+					r.Log.Error(err, "object is not type of ClusterAdmissionPolicy: %#v", policy)
+					queue.Add(ctrl.Request{})
 				}
-			}
-			policy.Status.PolicyStatus = policiesv1alpha2.ClusterAdmissionPolicyStatusPending
-			err = r.Reconciler.UpdateAdmissionPolicyStatus(context.Background(), policy)
-			if err != nil {
-				r.Log.Error(err, "cannot update status of policy "+policy.Name)
-				return []ctrl.Request{}
-			}
-			r.Log.Info("policy " + policy.Name + " pending")
-			return []ctrl.Request{
-				{
+				queue.Add(ctrl.Request{
 					NamespacedName: client.ObjectKey{
 						Name: policy.Spec.PolicyServer,
 					},
-				},
-			}
-		})).
-		Complete(r)
+				})
+			},
+			GenericFunc: func(e event.GenericEvent, queue workqueue.RateLimitingInterface) {
+				queue.Add(watchClusterAdmissionPolicy(r, e.Object))
+			},
```

Comment:
Out of curiosity, what is `GenericFunc` doing here? When is it called?

Reply:
The `controller-runtime` [documentation](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.1/pkg/event#GenericEvent) says:

```
GenericEvent is an event where the operation type is unknown (e.g. polling or event originating outside the cluster). GenericEvent should be generated by a source.Source and transformed into a reconcile.Request by an handler.EventHandler. 
```

This is the first time that I saw a generic event as well.

Reply:
Briefly looked on the internet and tests of controller-runtime and couldn't come up with examples of a GenericEvent. Given that they make come out of cluster, I think it's a good idea to register the function in the handler as we are doing it.
And I also like that `watchClusterAdmissionPolicy()` contains all the logic, and that the `GenericFunc` calls it, so it doesn't matter if it's a maskeraded create, update (at first I thought about moving the setting of `unscheduled` out of `watchClusterAdmissionPolicy()` and into `CreateFunc`, but I don't think it's a good idea if we keep `GenericFunc`).

Reply:
I can't find anything on the documentation. I ran a some tests, and everything seems to work fine without the `GenericFunc`. Although I'm not sure if there are some scenarios where it is called. I think we should be fine with `CreateFunc`, `UpdateFunc` and `DeleteFunc`.

Reply:
FYI: One more [piece of documentation](https://pkg.go.dev/sigs.k8s.io/controller-runtime@v0.11.1/pkg/handler#EventHandler) that I found in the `controller-runtime` lib

```
	// Generic is called in response to an event of an unknown type or a synthetic event triggered as a cron or
	// external trigger request - e.g. reconcile Autoscaling, or a Webhook.
```

I'm wondering if we set the callback to `nil` may cause a crash our controller to crash in some cases. Don`t you think? Maybe we can log when we receive this kind of event.

Reply:
I think is common to see only the one that controller-runtime users care about implemented (e.g. `CreateFunc` and `DeleteFunc` only).

The logic is https://github.com/kubernetes-sigs/controller-runtime/blob/f62a0f579d73fc2ea7b0aeed597521b83f9c3966/pkg/handler/eventhandler.go#L78-L104.

In any case, if you think this is me overreacting and you folks believe is better to keep it, I'm not strictly opposing to keep it, just that makes me a little nervous implementing something that we don't know if we need it or why.

Reply:
I agree that we should implement only the things that we know. Less code is better. My only concerns if we miss something. Because the previous [function forward any kind of event to our previous defined function](https://github.com/kubernetes-sigs/controller-runtime/blob/f62a0f579d73fc2ea7b0aeed597521b83f9c3966/pkg/handler/enqueue_mapped.go#L74-L77).

What I'm thinking to do is run the end-to-end tests in a version with no generic event implemented. If everything went fine, I'll remove the callback for this event type.

Reply:
Thank you @jvanz